File: code/javascript/algorithm_thinking_article-14.js
/**
 * Sliding window algorithm core code template
 */
export function solve(...args) {
  // Entry point for the course exercise: pass inputs directly.
  return lengthOfLongestSubstring(...args);
}

var lengthOfLongestSubstring = function(s) {
    let window = new Map();
    let left = 0, right = 0, res = 0;
    while (right < s.length) {
        let c = s[right];
        right++;
        window.set(c, (window.get(c) || 0) + 1);
        while (window.get(c) > 1) {
            let d = s[left];
            left++;
            window.set(d, window.get(d) - 1);
        }
        res = Math.max(res, right - left);
    }
    return res;
};

File: code/python/algorithm_thinking_article-14.py
"""Sliding window algorithm core code template"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().lengthOfLongestSubstring(*args, **kwargs)


class _Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        window = {}
        left = right = 0
        res = 0
        while right < len(s):
            c = s[right]
            right += 1
            window[c] = window.get(c, 0) + 1
            while window[c] > 1:
                d = s[left]
                left += 1
                window[d] -= 1
            res = max(res, right - left)
        return res
